### PR TITLE
Removed unnecessary MemoryMarshal.GetReference

### PR DESCRIPTION
**Pull request by lkts**
Fixes https://github.com/dotnet/corefx/issues/32099

There are few places where it breaks the logic, I marked them with comments and will highlight here in PR.

@JeremyKuhne @danmosemsft @jkotas

### CODE REVIEW

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs [hunk 1]
```diff
@@ -707,6 +707,7 @@ public static unsafe void ImplicitCast_ResultingSpanMatches(string s)
             ReadOnlySpan<char> span = s;
             Assert.Equal(s.Length, span.Length);
             fixed (char* stringPtr = s)
+            // removing MemoryMarshal breaks this test
```
Comment (lkts) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: here we try to compare pointer to empty string with pointer of empty span

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 2]
```diff
@@ -525,6 +525,11 @@ public override void Write(ReadOnlySpan<byte> buffer)
 
         internal void WriteCore(ReadOnlySpan<byte> buffer)
         {
+            if (buffer.IsEmpty)
```
Comment (lkts) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: nothing to write

**Comment by stephentoub** [comment]
Are all of these changes necessary to fix bugs? Or if not, is it possible any will refer performance on hot paths, due to the extra checks involved?

For ones that are fixing bugs, we should make sure we have tests that would have failed without these changes.

**Comment by stephentoub** [comment]
Changes under CoreLib should be made in the coreclr repo. I know it's a little confusing, but that code isn't necessarily used/compiled/tested in corefx; making the changes in coreclr will help validate it in CI.

**Comment by lkts** [comment]
@stephentoub this is just a cleanup, new code is equivalent except for empty span case and majority of places already have precondition checks for it (hence tests are passing)

**Comment by stephentoub** [comment]
> this is just a cleanup

Thanks (and for working on this). My concern still stands as to a) some of these changes being made in corefx instead of coreclr, and b) whether some of these should remain as they were in order to avoid the extra checks. In many/most cases, the extra checks won't matter and the simpler code you're changing it to is better, but I worry a few cases might regress perf.

**Comment by lkts** [comment]
@stephentoub I will port needed parts to coreclr if we decide to proceed with such changes. Regarding b) - that`s why i added comments in problematic places, other places are just result of "Find and Replace"

**Comment by stephentoub** [comment]
> other places are just result of "Find and Replace" :)

Those are the places I'm talking about.  Currently it's using:
https://github.com/dotnet/corefx/blob/cd861e116bba3079bc77b895b023995f8b5f92e7/src/Common/src/CoreLib/System/Runtime/InteropServices/MemoryMarshal.Fast.cs#L78
and your change makes it use:
https://github.com/dotnet/corefx/blob/cd861e116bba3079bc77b895b023995f8b5f92e7/src/Common/src/CoreLib/System/Span.Fast.cs#L161
which involves an extra branch to ensure that null is returned when the length is 0.  In the vast majority of uses, that extra branch won't matter.  But there are some places in these code bases where methods are used on such hots paths and in such tight loops that extra branches like that, and increases in code sizes, do matter.  It's quite possible that all of your changes fall into the "this doesn't matter" case, in which case great.  But that's my question.

**Comment by stephentoub** [comment]
> That totally makes sense, but how do i determine what to revert?

@JeremyKuhne, you opened #32099... what did you have in mind here?

@eerhardt just faced a case in his work where he saw upwards of a 20% dip in perf due to relying on the default use of GetPinnableReference rather than MemoryMarshal.GetReference, so a sweeping PR that switches from the latter to the former makes me nervous, even though I think it's unlikely to affect the majority of these uses.

@lkts, regardless, the changes under src/Common/src/CoreLib should be reverted in this PR.  Even if it's decided those changes should be made, they should be done via a PR to the coreclr repo rather than modifying the mirrored code in corefx.

Thanks.

**Comment by JeremyKuhne** [comment]
> you opened #32099... what did you have in mind here?

Consistency. Opened it based on @jkots feedback: https://github.com/dotnet/corefx/pull/32072#discussion_r214997132

**Comment by eerhardt** [comment]
> @eerhardt just faced a case in his work where he saw upwards of a 20% dip in perf due to relying on the default use of GetPinnableReference rather than MemoryMarshal.GetReference

See https://github.com/dotnet/machinelearning/pull/1229/commits/ec4c943cedf3ab535d69bac83062255c542827da where I am adding MemoryMarshal.GetReference to all ML.NET's calls to pin a span inside the CpuMath library.

On my machine the [K-Means benchmark](https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Benchmarks/KMeansAndLogisticRegressionBench.cs#L19) went from:

```
Using arrays:
           Method |     Mean |    Error |   StdDev | Extra Metric |
----------------- |---------:|---------:|---------:|-------------:|
 TrainKMeansAndLR | 869.8 ms | 83.13 ms | 95.73 ms |            - |


Using/pin Span with no MemoryMarshal.GetReference:
           Method |    Mean |    Error |   StdDev | Extra Metric |
----------------- |--------:|---------:|---------:|-------------:|
 TrainKMeansAndLR | 1.110 s | 0.2186 s | 0.2339 s |            - |


Pin Span with MemoryMarshal.GetReference:
           Method |     Mean |    Error |   StdDev | Extra Metric |
----------------- |---------:|---------:|---------:|-------------:|
 TrainKMeansAndLR | 868.9 ms | 87.16 ms | 100.4 ms |            - |
```

**Comment by lkts** [comment]
@stephentoub @eerhardt I see, thank you. I guess that PR and issue should be closed then and such changes should be done in scope of other changes to these files if it seems applicable.

**Comment by jkotas** [comment]
The array pinning has the extra checks for empty/null by default just like pinning for Span without 'MemoryMarshal.GetReference'. There is no fundamental reason why regular pinning of Span should be slower than regular pinning of Span. I agree that it would be good idea to understand the root cause of the performance difference in CpuMath library - since you do not see the performance regression with array pinning that has these null checks as well. E.g. You may be just seeing side-effects of cache or loop alignment that have nothing to do with pinning. Or you may be seeing JIT defficiency that we want to fix.

And since the array pinning has the extra checks for empty/null by default, folks are sometimes micro-optimizing it via patterns that avoid it. For example, if you  know that the array is not empty, you can do this:

```csharp
fixed (byte* p = &array[0])
{
   ...
}
```

This pattern tends to produce sligtly smaller and faster code than regular array pinning. It is similar to `MemoryMarshal.GetReference` micro-optimization for Span.

**Comment by stephentoub** [comment]
> For example, if you know that the array is not empty, you can do this

That's what Eric's array code was doing. The GetPinnableReference added checks not present in the array code.

**Comment by jkotas** [comment]
Even with everything mentioned here so far, the performance results in https://github.com/dotnet/corefx/pull/32669#issuecomment-429506746 do not work for me:

@eerhardt Could you please confirm:
- What was the exact version of the runtime you have run this benchmark on? 2.1 or 3.0? Note that 2.1 had number of Span-related performance bugs that are fixed in 3.0.
- For the "Using arrays" case, was the array pinned done as `fixed (void *p = array)` or as `fixed (void* p = &array[0])`?

**Comment by eerhardt** [comment]
>What was the exact version of the runtime you have run this benchmark on? Note that 2.1.x had number of Span-related performance bugs that are fixed in 3.0 preview builds.

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009622
  [Host]     : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  Job-RTCIWG : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
```

> For the "Using arrays" case, was the array pinned done as fixed (void *p = array) or as fixed (void* p = &array[0])?

All the existing CpuMath code pins the array with
`fixed (float* pa = &array[0])`

See https://github.com/dotnet/machinelearning/blob/7e9e4687284786ae241b2e8c8457a02686e639e9/src/Microsoft.ML.CpuMath/Sse.cs#L1120-L1134

**Comment by jkotas** [comment]
>  .NET Core 2.1.5

That can explain it. @eerhardt Could you please run the same on .NET Core 3.0 runtime ?

**Comment by stephentoub** [comment]
> Note that 2.1.x had number of Span-related performance bugs that are fixed in 3.0 preview builds.

@jkotas, which issues were those?

Should I infer from your comments that you believe it's safe to search/remove GetReference usage in exchange for the default GetPinnableReference, without further perf validation (in particular for the coreclr cases that were removed from this PR)?

**Comment by jkotas** [comment]
> @jkotas, which issues were those?

E.g. https://github.com/dotnet/coreclr/issues/18270

> Should I infer from your comments that you believe it's safe to search/remove GetReference usage in exchange for the default GetPinnableReference, without further perf validation

I agree with you that we should not do a blank search/remove without any perf-validation. This change should be split into multiple PRs:

1. We know about several cases where the GetReference pinning is likely causing real product bugs. There should be separate PRs with fixes for these, ideally with regression tests.
2. Go over GetReference uses for pining and flip the ones that are wrapping expensive operation (e.g. PInvoke) since there is unlikely to cause regression. Sanity check performance of a few randomly selected ones. Submit a PRs with this delta (it can be split into multiple PRs). 
3. For the remaining cases, go over them one-by-one and validate that there is no material perf regression when they are flipped.

My most recent comments were trying to suggest that the ML.NET results posted above do not represent the current state of .NET Core 3.0 well. The difference between the array pinning case (even the micro-optimized one with `array[0]`) and the regular Span pinning case should be non-existent or much smaller for ML.NET. And if it is not the case, there is likely RyuJIT bug that we want to understand and fix.

**Comment by eerhardt** [comment]
> @eerhardt Could you please run the same on .NET Core 3.0 runtime ?

I've run each commit 5 times using the latest 3.0 from `core-sdk`:

Arrays: (the "base" [commit](https://github.com/dotnet/machinelearning/commit/0b84350571fd0c272faa03541964c0d0617e0d76))
```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009670
  [Host]     : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT
  Job-KKDUIV : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT

BuildConfiguration=Release  Toolchain=netcoreapp3.0  MaxIterationCount=20
WarmupCount=1

           Method |     Mean |    Error |   StdDev | Extra Metric |
----------------- |---------:|---------:|---------:|-------------:|
 TrainKMeansAndLR | 760.0 ms | 46.96 ms | 46.12 ms |            - |
 TrainKMeansAndLR | 748.2 ms | 80.66 ms | 92.89 ms |            - |
 TrainKMeansAndLR | 766.3 ms | 95.77 ms | 106.4 ms |            - |
 TrainKMeansAndLR | 810.7 ms | 86.67 ms | 96.34 ms |            - |
 TrainKMeansAndLR | 794.4 ms | 69.48 ms | 77.23 ms |            - |
```

Pin Span (default): (Changing to use Spans [commit](https://github.com/dotnet/machinelearning/commit/9c7dc9cb4baec9d8dea837f6fa5ebc8dc76ea3cd))

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009670
  [Host]     : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT
  Job-LPAJUQ : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT

BuildConfiguration=Release  Toolchain=netcoreapp3.0  MaxIterationCount=20
WarmupCount=1

           Method |     Mean |    Error |   StdDev | Extra Metric |
----------------- |---------:|---------:|---------:|-------------:|
 TrainKMeansAndLR | 840.4 ms | 67.70 ms | 72.44 ms |            - |
 TrainKMeansAndLR | 830.0 ms | 81.18 ms | 93.49 ms |            - |
 TrainKMeansAndLR | 905.2 ms | 91.83 ms | 98.26 ms |            - |
 TrainKMeansAndLR | 902.4 ms | 117.0 ms | 134.8 ms |            - |
 TrainKMeansAndLR | 826.2 ms | 44.83 ms | 46.04 ms |            - |
```

Pin MemoryMarshal.GetReference(Span):(Changing to call MemoryMarshal [commit](https://github.com/dotnet/machinelearning/commit/ec4c943cedf3ab535d69bac83062255c542827da))

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009670
  [Host]     : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT
  Job-THNURV : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT

BuildConfiguration=Release  Toolchain=netcoreapp3.0  MaxIterationCount=20
WarmupCount=1

           Method |     Mean |    Error |   StdDev | Extra Metric |
----------------- |---------:|---------:|---------:|-------------:|
 TrainKMeansAndLR | 744.2 ms | 87.17 ms | 93.27 ms |            - |
 TrainKMeansAndLR | 778.3 ms | 64.35 ms | 71.53 ms |            - |
 TrainKMeansAndLR | 776.7 ms | 67.13 ms | 77.30 ms |            - |
 TrainKMeansAndLR | 735.3 ms | 57.70 ms | 61.74 ms |            - |
 TrainKMeansAndLR | 827.8 ms | 125.0 ms | 143.9 ms |            - |
```

You can see there is some variation run-to-run, but pinning the span in the "default" way is consistently higher.

**Comment by eerhardt** [comment]
Last week I also wrote up a micro benchmark, which I believe repros the core issue.

Here is the change with the benchmark, and a side-by-side "Span" overload to `DotProduct`
https://github.com/eerhardt/machinelearning/commit/56b849c6d54bfd9e73d8bec8c119660d3e5b6dc1

Running the above benchmark produces:

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009670
  [Host]     : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT
  Job-OWUCOK : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT

BuildConfiguration=Release  Toolchain=netcoreapp3.0  MaxIterationCount=20
WarmupCount=1

               Method |     Mean |     Error |    StdDev | Extra Metric |
--------------------- |---------:|----------:|----------:|-------------:|
   TestDotProductSpan | 61.28 ns | 1.5821 ns | 1.8220 ns |            - |
 TestDotProductNoSpan | 44.95 ns | 0.8567 ns | 0.7594 ns |            - |
```

And here is the change to call `MemoryMarshal.GetReference`: 
https://github.com/eerhardt/machinelearning/commit/5197bfabb1a93047ff0eff9a6dddb521ee1c8281

Running the benchmark again:

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009670
  [Host]     : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT
  Job-EHBIXH : .NET Core 3.0.0-preview1-27009-01 (CoreCLR 4.6.27008.04, CoreFX 4.6.27008.04), 64bit RyuJIT

BuildConfiguration=Release  Toolchain=netcoreapp3.0  MaxIterationCount=20
WarmupCount=1

               Method |     Mean |     Error |    StdDev | Extra Metric |
--------------------- |---------:|----------:|----------:|-------------:|
   TestDotProductSpan | 47.93 ns | 0.9330 ns | 0.8727 ns |            - |
 TestDotProductNoSpan | 44.85 ns | 1.3614 ns | 1.5132 ns |            - |
```

**Comment by jkotas** [comment]
Thanks. I have tweaked GetPinnableReference implementation in https://github.com/dotnet/coreclr/pull/20428 to make it much closer to pinning using `MemoryMarshal.GetReference`.

**Comment by lkts** [comment]
Thanks everyone for productive discussion! Glad that improvement possibility was  discovered here. As a result I suggest updating https://github.com/dotnet/corefx/issues/32099 with steps provided by @jkotas and other relevant information (or creating new issue). Code changes here do not make sense anymore so closing it.

**Pull request closed without merging**